### ISSUE TITLE
literatures for activity_regularizer

### ISSUE THREAD

**Issue by keunwoochoi** [fchollet/keras#1618]
I am looking for methods that motivate activity to be sparse, and thinking `activity_regularizer` sould be one of the methods. However I failed to find any literature on it in Google. 
Of course I must be using wrong keywords, but does anyone know literatures to get some background of activity_regularizer?

**Comment by ymcui** [comment]
Here is the source code of `WeightRegularizer()` and `ActivityRegularizer()`, I think it is obvious to see the difference. 
https://github.com/fchollet/keras/blob/master/keras/regularizers.py#L19
```
class WeightRegularizer(Regularizer):
    def __init__(self, l1=0., l2=0.):
        self.l1 = l1
        self.l2 = l2

    def set_param(self, p):
        self.p = p

    def __call__(self, loss):
        loss += K.sum(K.abs(self.p)) * self.l1
        loss += K.sum(K.square(self.p)) * self.l2
        return loss

    def get_config(self):
        return {"name": self.__class__.__name__,
                "l1": self.l1,
                "l2": self.l2}

class ActivityRegularizer(Regularizer):
    def __init__(self, l1=0., l2=0.):
        self.l1 = l1
        self.l2 = l2

    def set_layer(self, layer):
        self.layer = layer

    def __call__(self, loss):
        output = self.layer.get_output(True)
        loss += self.l1 * K.sum(K.mean(K.abs(output), axis=0))
        loss += self.l2 * K.sum(K.mean(K.square(output), axis=0))
        return loss

    def get_config(self):
        return {"name": self.__class__.__name__,
                "l1": self.l1,
                "l2": self.l2}
```

**Comment by keunwoochoi** [comment]
Hi, thank you but I realise how they work and how they are different. I was looking for some research articles about their effects e.g. sparsity, noise model, etc.

**Comment by phreeza** [comment]
@article{ng2011sparse,
  title={Sparse autoencoder},
  author={Ng, Andrew},
  journal={CS294A Lecture notes},
  volume={72},
  pages={1--19},
  year={2011}
}

http://web.stanford.edu/class/cs294a/sae/sparseAutoencoderNotes.pdf

**Comment by keunwoochoi** [comment]
Seems like it's what I need, thanks.

**Comment by tboquet** [comment]
http://www.deeplearningbook.org/contents/autoencoders.html

**Comment by keunwoochoi** [comment]
Thanks for another good material. 
For someone who's interested, they propose different types of sparsity regularisation. 
Andrew Ng's lecture note uses KL divergence between desired sparsity parameter and what we have, and put it into the loss function. Deep learning book introduces a sparsity regularisation using the thing keras has as an `activity_l1`, i.e. adding absolute values of activities into the loss.

**Comment by fchollet** [comment]
> Deep learning book introduces

Which book?

**Comment by keunwoochoi** [comment]
[This book](http://www.deeplearningbook.org/contents/autoencoders.html) that @tboquet shared above, and Ian Goodfellow et al. are writing.
